--- node_modules/csurf/index.js	1985-10-26 01:15:00.000000000 -0700
+++ node_modules/csurf/index.js	2019-08-31 00:06:51.000000000 -0700
@@ -65,6 +65,9 @@
   // generate lookup
   var ignoreMethod = getIgnoredMethods(ignoreMethods)
 
+  // get csrf invalid event emitter
+  var invalidEventEmitter = opts.eventEmitter;
+
   return function csrf (req, res, next) {
     // validate the configuration against request
     if (!verifyConfiguration(req, sessionKey, cookie)) {
@@ -109,12 +112,16 @@
 
     // verify the incoming token
     if (!ignoreMethod[req.method] && !tokens.verify(secret, value(req))) {
-      return next(createError(403, 'invalid csrf token', {
-        code: 'EBADCSRFTOKEN'
-      }))
+      if (invalidEventEmitter && typeof invalidEventEmitter.emit === 'function') {
+        invalidEventEmitter.emit('csrfInvalidated', req, res);
+      } else {
+        return next(createError(403, 'invalid csrf token', {
+          code: 'EBADCSRFTOKEN'
+        }))
+      }
+    } else {
+      next();
     }
-
-    next()
   }
 }
 
