--- old/contrail-webui-third-party/node_modules/easysoap/lib/wsdl.js	2014-11-10 03:29:31.883488143 -0800
+++ new/contrail-webui-third-party/node_modules/easysoap/lib/wsdl.js	2014-11-10 03:30:24.611488143 -0800
@@ -242,7 +242,8 @@
                 else {
 
                     params.path = params.wsdl;
-
+                    resolve(fs.readFileSync(params.path, 'UTF-8'));
+                    /*
                     request.get(params, opts)
                         .done(function(res) {
                             fs.writeFile(fullPath, res.body, function(err) {
@@ -256,6 +257,7 @@
                         function(err) {
                             reject(err);
                         });
+                    */
                 }
             });
         }
@@ -459,4 +461,4 @@
     };
 
 
-})();
\ No newline at end of file
+})();
--- old/contrail-webui-third-party/node_modules/easysoap/lib/request/request.tpl	2014-11-10 03:29:31.883488143 -0800
+++ new/contrail-webui-third-party/node_modules/easysoap/lib/request/request.tpl	2014-11-10 03:31:40.381488143 -0800
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <SOAP-ENV:Envelope
-    xmlns:SOAP-ENV="<%= envelope.soap_env%>"
+    xmlns:SOAP-ENV="<%= envelope.soap_env%>" xmlns:xsd='http://www.w3.org/2001/XMLSchema' xmlns:xsi='http://www.w3.org/2001/XMLSchema-instance'
     <% if (envelope.namespaces !== null) { %>
         <% _.each(envelope.namespaces, function(namespace) { %>
             <% if (namespace.full !== void 0) { %>
@@ -35,4 +35,4 @@
         <% } %>
     </SOAP-ENV:Body>
 
-</SOAP-ENV:Envelope>
\ No newline at end of file
+</SOAP-ENV:Envelope>
