--- jdorn-jsoneditor.js	2016-09-16 15:52:34.000000000 +0530
+++ jdorn-jsoneditor.js	2016-11-29 15:38:05.000000000 +0530
@@ -125,11 +125,11 @@
 // Array.isArray polyfill
 // From MDN
 (function() {
-	if(!Array.isArray) {
-	  Array.isArray = function(arg) {
-		return Object.prototype.toString.call(arg) === '[object Array]';
-	  };
-	}
+    if(!Array.isArray) {
+      Array.isArray = function(arg) {
+        return Object.prototype.toString.call(arg) === '[object Array]';
+      };
+    }
 }());
 /**
  * Taken from jQuery 2.1.3
@@ -215,24 +215,95 @@
   if (!(element instanceof Element)) {
     throw new Error('element should be an instance of Element');
   }
+  objectPropertiesStack = [];
+  propertyDeleteCount = 0;
+  directLevelKey = [];
+  removedKeyStack = [];
+  enableInputField = true
+  addPropertyFlag = false;
+  disableCheckBoxTrash = true;
+  removedKeyStack = [];
+  requiredKey = [];
+  disableTashIconInArray = true;
+  blockFirstOptions = true;
   options = $extend({},JSONEditor.defaults.options,options||{});
   this.element = element;
   this.options = options;
+  clickAddItem = false;
   this.init();
 };
 JSONEditor.prototype = {
   // necessary since we remove the ctor property by doing a literal assignment. Without this
   // the $isplainobject function will think that this is a plain object.
   constructor: JSONEditor,
+  addAdditionalFlagInProp :function(schema, order) {
+      var self = this;
+      for (var i in schema) {
+          if (typeof schema[i] === 'object' || i === 'properties') {
+                 self.addAdditionalFlagInProp(schema[i], order);
+          }else{
+              if(i !== 'properties'){
+                  if(!schema.hasOwnProperty('collapse')){
+                      if(order === undefined){
+                          schema.collapse = true;
+                      }else{
+                          schema.collapse = false;
+                      }
+                  }
+               }
+              if(schema[i] === 'boolean'){
+                  schema.format = "checkbox";
+              }
+              if(schema[i] === 'string' || schema[i] === 'number'){
+                  schema.readOnly = false;
+              }
+          }
+      }
+    return schema;
+  },
+  getAllrequiredFields: function(properties){
+      var self = this;
+      for(var i in properties){
+          if(properties[i].type == 'object'){
+              if(properties[i].required !== undefined){
+                  if(properties[i].required === 'required' || properties[i].required === 'true'){
+                      requiredKey.push(i);
+                  }
+              }
+              if(properties[i].properties !== undefined){
+                  self.getAllrequiredFields(properties[i].properties);
+              }
+          }else if(properties[i].type == 'array'){
+              if(properties[i].required !== undefined){
+                  if(properties[i].required === 'required' || properties[i].required === 'true'){
+                      requiredKey.push(i);
+                  }
+              }
+              if(properties[i].items !== undefined){
+                  if(properties[i].items.properties !== undefined){
+                      self.getAllrequiredFields(properties[i].items.properties);
+                  }
+              }
+          }else{
+              if(properties[i].required !== undefined){
+                  if(properties[i].required === 'required' || properties[i].required === 'true'){
+                      requiredKey.push(i);
+                  }
+              }
+          }
+      }
+  },
   init: function() {
-    var self = this;
-    
+    var self = this,order;
     this.ready = false;
-
     var theme_class = JSONEditor.defaults.themes[this.options.theme || JSONEditor.defaults.theme];
     if(!theme_class) throw "Unknown theme " + (this.options.theme || JSONEditor.defaults.theme);
-    
+    this.addAdditionalFlagInProp(this.options.schema, order);
+    delete this.options.schema.collapse;
+    this.options.schema.properties[Object.keys(this.options.schema.properties)[0]].collapse = false;
     this.schema = this.options.schema;
+    var properties = this.schema.properties[Object.keys(this.schema.properties)[0]].properties;
+    this.getAllrequiredFields(properties);
     this.theme = new theme_class();
     this.template = this.options.template;
     this.refs = this.options.refs || {};
@@ -314,7 +385,6 @@
   destroy: function() {
     if(this.destroyed) return;
     if(!this.ready) return;
-    
     this.schema = null;
     this.options = null;
     this.root.destroy();
@@ -328,14 +398,12 @@
     this.__data = null;
     this.ready = false;
     this.element.innerHTML = '';
-    
     this.destroyed = true;
   },
   on: function(event, callback) {
     this.callbacks = this.callbacks || {};
     this.callbacks[event] = this.callbacks[event] || [];
     this.callbacks[event].push(callback);
-    
     return this;
   },
   off: function(event, callback) {
@@ -359,7 +427,6 @@
     else {
       this.callbacks = {};
     }
-    
     return this;
   },
   trigger: function(event) {
@@ -368,7 +435,6 @@
         this.callbacks[event][i]();
       }
     }
-    
     return this;
   },
   setOption: function(option, value) {
@@ -380,7 +446,6 @@
     else {
       throw "Option "+option+" must be set during instantiation and cannot be changed later";
     }
-    
     return this;
   },
   getEditorClass: function(schema) {
@@ -412,9 +477,9 @@
     
     if(this.firing_change) return;
     this.firing_change = true;
-    
+
     var self = this;
-    
+
     window.requestAnimationFrame(function() {
       self.firing_change = false;
       if(!self.ready) return;
@@ -1432,6 +1497,7 @@
   setContainer: function(container) {
     this.container = container;
     if(this.schema.id) this.container.setAttribute('data-schemaid',this.schema.id);
+    if(this.schema.type === null) this.schema.type = 'string';
     if(this.schema.type && typeof this.schema.type === "string") this.container.setAttribute('data-schematype',this.schema.type);
     this.container.setAttribute('data-schemapath',this.path);
   },
@@ -1754,7 +1820,13 @@
     return null;
   },
   getTitle: function() {
-    return this.schema.title || this.key;
+      var splitedKey = this.key.split('_'); var strStack = [];
+      for(var i = 0; i < splitedKey.length; i++){
+          var captilizeStr = splitedKey[i].charAt(0).toUpperCase() + splitedKey[i].slice(1);
+          strStack.push(captilizeStr);
+      }
+      return strStack.join(' ');
+    //return this.schema.title || this.key;
   },
   enable: function() {
     this.disabled = false;
@@ -1868,7 +1940,9 @@
   },
   setValue: function(value,initial,from_template) {
     var self = this;
-    
+    if(addPropertyFlag){
+        value = null;
+    }
     if(this.template && !from_template) {
       return;
     }
@@ -1924,8 +1998,11 @@
   build: function() {
     var self = this, i;
     if(!this.options.compact) this.header = this.label = this.theme.getFormInputLabel(this.getTitle());
-    if(this.schema.description) this.description = this.theme.getFormInputDescription(this.schema.description);
-
+    if(this.schema.description !== undefined){
+        if(!this.schema.description){
+            this.description = this.theme.getFormInputDescription(this.schema.description);
+        }
+    }
     this.format = this.schema.format;
     if(!this.format && this.schema.media && this.schema.media.type) {
       this.format = this.schema.media.type.replace(/(^(application|text)\/(x-)?(script\.)?)|(-source$)/g,'');
@@ -2040,7 +2117,22 @@
     else {
       if(this.options.input_width) this.input.style.width = this.options.input_width;
     }
-
+  //set readOnly true
+    if(this.schema.type == 'string' || this.schema.type == 'number'){
+            if(this.options.jsoneditor.options.unDeletableProperty.indexOf(this.label.innerHTML) >= 0){
+                if(enableInputField){
+                    this.schema.readOnly = true;
+                    this.hideIconForProp = false; 
+                }else{
+                    this.schema.readOnly = false;
+                    this.hideIconForProp = true;
+                }
+                
+            }else{
+                this.schema.readOnly = false;
+                this.hideIconForProp = true;  
+            }
+        }
     if(this.schema.readOnly || this.schema.readonly || this.schema.template) {
       this.always_disabled = true;
       this.input.disabled = true;
@@ -2108,8 +2200,15 @@
     }
 
     if(this.format) this.input.setAttribute('data-schemaformat',this.format);
-
-    this.control = this.theme.getFormControl(this.label, this.input, this.description);
+    if(this.schema.description){
+        this.input.setAttribute('title',this.schema.description);
+    }
+    if(this.schema.required == 'required' || this.schema.required == 'true'){
+        var icon = document.createElement('i');
+        icon.className = 'fa fa-asterisk jsonEditor-form-asterisk';
+        this.label.appendChild(icon);
+    }
+    this.control = this.theme.getFormControl(this.label, this.input, this.description,this.hideIconForProp);
     this.container.appendChild(this.control);
 
     // Any special formatting that needs to happen after the input is added to the dom
@@ -2482,13 +2581,34 @@
     else {
       container = document.createElement('div');
       $each(this.property_order, function(i,key) {
+        var size = 12;
         var editor = self.editors[key];
+        var element = editor.path.split('.');
+        if(element.length == 3){
+            if(editor.schema.type == 'number' || editor.schema.type == 'string'){
+                if(editor.schema.propertyOrder == 1){
+                    editor.container.setAttribute('style','width:97.4%;');
+                    editor.container.firstChild.setAttribute('style','width:49.4%;');
+                }
+                if(key == 'parent_href' || key == 'href'){
+                    editor.container.firstChild.lastChild.firstChild.setAttribute('style','width:97.6%;height:28px;');
+                    size = 12;
+                }else{
+                    size = 6;
+                }
+            }
+            if(editor.schema.type == 'boolean'){
+                //editor.container.setAttribute('style','height:45px;');
+                //editor.container.firstChild.setAttribute('style','margin-top:13px;');
+                size = 6;
+            }
+        }
         if(editor.property_removed) return;
         var row = self.theme.getGridRow();
         container.appendChild(row);
 
         if(editor.options.hidden) editor.container.style.display = 'none';
-        else self.theme.setGridColumnSize(editor.container,12);
+        else self.theme.setGridColumnSize(editor.container,size);
         row.appendChild(editor.container);
       });
     }
@@ -2591,8 +2711,9 @@
   },
   build: function() {
     var self = this;
+    objectDelete = false;
+    trashIconHide = undefined;
 
-    // If the object should be rendered as a table row
     if(this.options.table_row) {
       this.editor_holder = this.container;
       $each(this.editors, function(key,editor) {
@@ -2619,16 +2740,16 @@
     // If the object should be rendered as a div
     else {
       this.header = document.createElement('span');
-      this.header.textContent = this.getTitle();
-      this.title = this.theme.getHeader(this.header);
+      //this.header.textContent = this.getTitle();
+      this.title = this.theme.getHeader('');
       this.container.appendChild(this.title);
       this.container.style.position = 'relative';
 
       // Edit JSON modal
       this.editjson_holder = this.theme.getModal();
       this.editjson_textarea = this.theme.getTextareaInput();
-      this.editjson_textarea.style.height = '170px';
-      this.editjson_textarea.style.width = '300px';
+      this.editjson_textarea.style.height = '300px';
+      this.editjson_textarea.style.width = '600px';
       this.editjson_textarea.style.display = 'block';
       this.editjson_save = this.getButton('Save','save','Save');
       this.editjson_save.addEventListener('click',function(e) {
@@ -2648,6 +2769,7 @@
 
       // Manage Properties modal
       this.addproperty_holder = this.theme.getModal();
+      this.addproperty_holder.className = 'addPropertyHolder';
       this.addproperty_list = document.createElement('div');
       this.addproperty_list.style.width = '295px';
       this.addproperty_list.style.maxHeight = '160px';
@@ -2656,13 +2778,14 @@
       this.addproperty_list.style.overflowX = 'hidden';
       this.addproperty_list.style.paddingLeft = '5px';
       this.addproperty_list.setAttribute('class', 'property-selector');
-      this.addproperty_add = this.getButton('add','add','add');
+      this.addproperty_add = this.getButton('Add','','');
       this.addproperty_input = this.theme.getFormInputField('text');
       this.addproperty_input.setAttribute('placeholder','Property name...');
       this.addproperty_input.style.width = '220px';
       this.addproperty_input.style.marginBottom = '0';
       this.addproperty_input.style.display = 'inline-block';
       this.addproperty_add.addEventListener('click',function(e) {
+         // this.types = ['string','number','integer','boolean','null'];
         e.preventDefault();
         e.stopPropagation();
         if(self.addproperty_input.value) {
@@ -2677,6 +2800,7 @@
           }
           self.onChange(true);
         }
+        self.hideAddProperty();
       });
       this.addproperty_holder.appendChild(this.addproperty_list);
       this.addproperty_holder.appendChild(this.addproperty_input);
@@ -2688,10 +2812,18 @@
 
       // Description
       if(this.schema.description) {
-        this.description = this.theme.getDescription(this.schema.description);
-        this.container.appendChild(this.description);
+        //this.description = this.theme.getDescription(this.schema.description);
+        //this.container.appendChild(this.description);
+      }
+      if(this.schema.title !== undefined){
+          if(this.schema.title.search('item') == 0){
+              if(this.schema.required !== undefined){
+                  if(this.schema.required.length > 1){
+                      trashIconHide = true;
+                  }
+              }
+          }
       }
-
       // Validation error placeholder area
       this.error_holder = document.createElement('div');
       this.container.appendChild(this.error_holder);
@@ -2707,36 +2839,146 @@
       $each(this.editors, function(key,editor) {
         var holder = self.theme.getGridColumn();
         self.row_container.appendChild(holder);
-
         editor.setContainer(holder);
-        editor.build();
+        editor.build(trashIconHide);
         editor.postBuild();
       });
+      this.headerText = document.createElement('span');
+      this.headerText.textContent = this.getTitle();
 
-      // Control buttons
       this.title_controls = this.theme.getHeaderButtonHolder();
       this.editjson_controls = this.theme.getHeaderButtonHolder();
       this.addproperty_controls = this.theme.getHeaderButtonHolder();
       this.title.appendChild(this.title_controls);
+      this.title.appendChild(this.headerText);
       this.title.appendChild(this.editjson_controls);
-      this.title.appendChild(this.addproperty_controls);
+      if(this.schema.description){
+          this.title.lastChild.previousSibling.setAttribute('title',this.schema.description);
+      }
+      if(this.schema.required == 'required' || this.schema.required == 'true'){
+          var icon = document.createElement('i');
+          icon.className = 'fa fa-asterisk jsonEditor-form-obj-asterisk';
+          this.title.lastChild.previousSibling.appendChild(icon);
+      }
+      if(addPropertyFlag){
+          this.title.appendChild(this.addproperty_controls);
+      }
+      if(this.options.path !== undefined){
+          var elePath = this.options.path.split('.');
+          if(elePath.length > 2){
+             var index = true;
+                  var trashObj = document.createElement('button');
+                  trashObj.type = 'button';
+                  trashObj.style.width = '20px'
+                  trashObj.className = 'btn btn-mini json-editor-btn-trash deleteChildLevel';
+                  var objInnerText = document.createElement('i');
+                  objInnerText.className = 'fa fa-trash-o';
+                  trashObj.appendChild(objInnerText);
+                  trashObj.addEventListener('click',function(e) {
+                      e.preventDefault();
+                      e.stopPropagation();
+                      var text = jsoneditor.root.parseJsonKeyUpperToLower(this.parentElement.firstChild.nextElementSibling.textContent);
+                      if(text == 'virtual_dNS_record_data'){
+                          text = 'virtual_DNS_record_data';
+                      }
+                      if(requiredKey.indexOf(text) === -1){
+                          objectDelete = true;
+                          var attrValue = this.parentElement.parentElement.getAttribute('data-schemapath').split('.');
+                          attrValue.splice(0,1);
+                          for(var k = 0; k < attrValue.length; k++){
+                              if(isNumber(attrValue[k])){
+                                  index = false;
+                              }
+                          }
+                          if(index){
+                              $(this).closest('.row-fluid')[0].remove();
+                          }else{
+                              this.parentElement.parentElement.remove();
+                          }
+                          jsoneditor.root.removeObjectProperty(attrValue,true);
+                          jsoneditor.root.onChange(true);
+                          objectDelete = false;
+                      }else{
+                          var errorHolder = $("#config-error-msg-container");
+                          errorHolder.empty();
+                          errorHolder.text('We Can Not Delete The Required Filed - '+ this.parentElement.firstChild.nextElementSibling.textContent);
+                          $('.modal-body').scrollTop(0);
+                          var errorContainer = $("#config-error-container");
+                          errorContainer.fadeIn(500);
+                          $("#remove-error-popup").on('click',function(){
+                              $("#config-error-container").css('display','none');
+                          });
+                      }
+                    });
+                  this.title.appendChild(trashObj);
+          }
+      }
 
       // Show/Hide button
-      this.collapsed = false;
-      this.toggle_button = this.getButton('','collapse','Collapse');
-      this.title_controls.appendChild(this.toggle_button);
+      var schemaPath = self.editor_holder.parentElement.getAttribute('data-schemapath').split('.');
+      if(schemaPath.length === 3){
+          self.editor_holder.parentElement.firstChild.setAttribute('style','border-bottom:1px solid #eeeeee;');
+      }else if(schemaPath.length >= 4){
+          self.editor_holder.parentElement.lastChild.firstChild.setAttribute('style','border-left: 1px dotted #e5e5e5;margin-left: 9px;');
+      }else if(schemaPath.length == 2){
+          self.editor_holder.parentElement.setAttribute('style','position: relative;padding-right:0px;padding-left:11px;');
+      }
+      if(this.schema.collapse){
+          if(this.schema.title !== undefined){
+              var title = this.schema.title.split('');
+              if(title[0] == '[' && clickAddItem){
+                  this.toggle_button = this.getButton('','collapse','Collapse');
+                  this.title_controls.appendChild(this.toggle_button);
+                  self.editor_holder.style.display = '';
+                  this.collapsed = false;
+              }else{
+                  this.toggle_button = this.getButton('','expand','Expand');
+                  this.title_controls.appendChild(this.toggle_button);
+                  self.editor_holder.style.display = 'none';
+                  this.collapsed = true;
+              }
+          }else{
+              this.toggle_button = this.getButton('','expand','Expand');
+              this.title_controls.appendChild(this.toggle_button);
+              self.editor_holder.style.display = 'none';
+              this.collapsed = true;
+          }
+      }else{
+          this.toggle_button = this.getButton('','collapse','Collapse');
+          this.title_controls.appendChild(this.toggle_button);
+          self.editor_holder.style.display = '';
+          this.collapsed = false;
+      }
+      this.title.addEventListener('click',function(e) {
+          e.preventDefault();
+          e.stopPropagation();
+          if(self.collapsed) {
+            $(self.editor_holder).slideDown(700);
+            self.collapsed = false;
+            self.setButtonText(self.toggle_button,'','collapse','Collapse');
+            this.firstChild.firstChild.setAttribute('style','margin-top:-8px;');
+          }
+          else {
+            $(self.editor_holder).slideUp(700);
+            self.collapsed = true;
+            self.setButtonText(self.toggle_button,'','expand','Expand');
+            this.firstChild.firstChild.setAttribute('style','margin-top:0px;');
+          }
+      });
       this.toggle_button.addEventListener('click',function(e) {
         e.preventDefault();
         e.stopPropagation();
         if(self.collapsed) {
-          self.editor_holder.style.display = '';
+          $(self.editor_holder).slideDown(700);
           self.collapsed = false;
           self.setButtonText(self.toggle_button,'','collapse','Collapse');
+          this.setAttribute('style','margin-top:-8px;');
         }
         else {
-          self.editor_holder.style.display = 'none';
+          $(self.editor_holder).slideUp(700);
           self.collapsed = true;
           self.setButtonText(self.toggle_button,'','expand','Expand');
+          this.setAttribute('style','margin-top:0px;');
         }
       });
 
@@ -2754,8 +2996,9 @@
       }
 
       // Edit JSON Button
-      this.editjson_button = this.getButton('JSON','edit','Edit JSON');
+      this.editjson_button = this.getButton('','edit','Edit JSON');
       this.editjson_button.addEventListener('click',function(e) {
+        //disableTashIconInArray = true;
         e.preventDefault();
         e.stopPropagation();
         self.toggleEditJSON();
@@ -2770,16 +3013,34 @@
       else if(this.jsoneditor.options.disable_edit_json) {
         this.editjson_button.style.display = 'none';
       }
-
+      $(document).ready(function () {
+          $('.addPropertyHolder').hide();
+      });
+      $('.addObjectPropperty').on('click', function () {
+      });
+     $(document).mouseup(function (e) {
+          var popup = $(".addPropertyHolder");
+          if (!$('.addObjectPropperty').is(e.target.parentElement) && !popup.is(e.target) && popup.has(e.target).length == 0) {
+              popup.hide();
+              jsoneditor.root.addingProperty = false;
+              self.enable();
+          }
+      });
       // Object Properties Button
-      this.addproperty_button = this.getButton('Properties','edit','Object Properties');
+      this.addproperty_button = this.getButton('','add','Object Properties');
+      this.addproperty_button.className='btn btn-mini json-editor-btn-plus addObjectPropperty';
       this.addproperty_button.addEventListener('click',function(e) {
         e.preventDefault();
         e.stopPropagation();
-        self.toggleAddProperty();
-      });
-      this.addproperty_controls.appendChild(this.addproperty_button);
-      this.addproperty_controls.appendChild(this.addproperty_holder);
+        disableCheckBoxTrash = false;
+        var siblingText = this.parentElement.previousSibling.previousSibling.innerHTML;
+        self.toggleAddProperty(siblingText);
+        disableCheckBoxTrash = true;
+      });
+      if(addPropertyFlag){
+          this.addproperty_controls.appendChild(this.addproperty_button);
+          this.addproperty_controls.appendChild(this.addproperty_holder); 
+      }
       this.refreshAddProperties();
     }
 
@@ -2817,6 +3078,14 @@
     this.editjson_button.disabled = false;
     this.editing_json = true;
   },
+  parseJsonKeyUpperToLower: function(key){
+      var splitedKey = key.split(' '); var strStack = [];
+      for(var i = 0; i < splitedKey.length; i++){
+          var captilizeStr = splitedKey[i].charAt(0).toLowerCase() + splitedKey[i].slice(1);
+          strStack.push(captilizeStr);
+      }
+      return strStack.join('_');
+  },
   hideEditJSON: function() {
     if(!this.editjson_holder) return;
     if(!this.editing_json) return;
@@ -2864,10 +3133,24 @@
   addPropertyCheckbox: function(key) {
     var self = this;
     var checkbox, label, labelText, control;
-
+    if(this.options.path !== undefined){
+        var parentPath = this.options.path.split('.');
+        var parentObj = parentPath[2];
+        var childObj = parentPath[parentPath.length - 1];
+    }
+    if(removedKeyStack.length > 0){
+        for(var i=0; i < removedKeyStack.length; i++){
+            var keyValue = removedKeyStack[i].split('.');
+            var delParentKey = keyValue[2];
+            var delChildKey = keyValue[0];
+            if(delParentKey == parentObj && delChildKey == childObj){
+                key = null;
+            }
+        }
+    }
     checkbox = self.theme.getCheckbox();
     checkbox.style.width = 'auto';
-
+   if(key !== null){
     if (this.schema.properties[key] && this.schema.properties[key].title)
       labelText = this.schema.properties[key].title;
     else
@@ -2884,19 +3167,21 @@
 
     checkbox.checked = key in this.editors;
     checkbox.addEventListener('change',function() {
-      if(checkbox.checked) {
+       disableCheckBoxTrash = false;
+       if(checkbox.checked) {
         self.addObjectProperty(key);
-      }
-      else {
+       }
+       else {
         self.removeObjectProperty(key);
-      }
-      self.onChange(true);
+       }
+       self.onChange(true);
     });
     self.addproperty_checkboxes[key] = checkbox;
 
     return checkbox;
+    }
   },
-  showAddProperty: function() {
+  showAddProperty: function(node) {
     if(!this.addproperty_holder) return;
     this.hideEditJSON();
 
@@ -2911,7 +3196,8 @@
     this.adding_property = true;
     this.addproperty_button.disabled = false;
     this.addproperty_holder.style.display = '';
-    this.refreshAddProperties();
+    this.refreshAddProperties(node);
+    jsoneditor.root.addingProperty = true;
   },
   hideAddProperty: function() {
     if(!this.addproperty_holder) return;
@@ -2922,17 +3208,63 @@
 
     this.adding_property = false;
   },
-  toggleAddProperty: function() {
+  toggleAddProperty: function(node) {
+    if(!jsoneditor.root.addingProperty){
+        this.adding_property = false;
+    }
     if(this.adding_property) this.hideAddProperty();
-    else this.showAddProperty();
+    else this.showAddProperty(node);
   },
-  removeObjectProperty: function(property) {
-    if(this.editors[property]) {
-      this.editors[property].unregister();
-      delete this.editors[property];
-
-      this.refreshValue();
-      this.layoutEditors();
+  hideAddPropertyByContainer: function(node) {
+      this.hideAddProperty();
+  },
+  removeObjectProperty: function(property,flag) {
+      var self = this;
+      currentObj = undefined;
+      arrayParentKey = undefined;
+      delFlag = false;
+      childLevelObj = undefined;
+      keyHierarchy = [];
+      var childKeyDelete = true;
+      if(flag){
+         var topObjLevel = property.join(';');
+         var lastlevel = property.splice(property.length - 1, 1);
+         parentLevelObj = property[property.length - 1];
+         for(var i = 0; i < property.length; i++){
+             keyHierarchy.push(property[i]);
+             if(isNumber(property[i])){
+                 childKeyDelete = false;
+             }
+         }
+         keyHierarchy.shift();
+         keyHierarchy.shift();
+         childLevelObj = property;
+         directLevelKey =[];
+         directLevelKey.push(lastlevel[0]);
+         parentObjCount = 'empty';
+         arrayParentKey = property[1];
+         var addObj = {};
+         if(property.length > 2){
+             var parentObjCount = property[2];
+         }
+         addObj.parentKey = parentLevelObj;
+         addObj.childKey = lastlevel[0];
+         objectPropertiesStack.push(addObj);
+         if(property.length === 1){
+             removedKeyStack.push('directChild'+'.'+parentObjCount+'.'+'noParent'+'.'+objectDelete+'.'+lastlevel[0]);
+         }else{
+             removedKeyStack.push(property[property.length - 1]+'.'+parentObjCount+'.'+property[1]+'.'+objectDelete+'.'+lastlevel[0]);
+         }
+         this.refreshValue();
+         this.layoutEditors();
+    }else{
+        if(this.editors[property]) {
+            this.editors[property].unregister();
+            delete this.editors[property];
+
+            this.refreshValue();
+            this.layoutEditors();
+          } 
     }
   },
   addObjectProperty: function(name, prebuild_only) {
@@ -3010,6 +3342,7 @@
     this._super();
   },
   getValue: function() {
+      var self = this;
     var result = this._super();
     if(this.jsoneditor.options.remove_empty_properties || this.options.remove_empty_properties) {
       for(var i in result) {
@@ -3018,7 +3351,190 @@
         }
       }
     }
-    return result;
+    if(directLevelKey.length > 0){
+        if(result[directLevelKey[0]] !== undefined){
+            if(childLevelObj.length > 1){
+                var updatedResult = self.deleteProperty(result,directLevelKey[0],result);
+                directLevelKey =[];
+                return updatedResult;
+            }else{
+                //delete result[directLevelKey[0]];
+                if(result[directLevelKey[0]].constructor === Array){
+                    result[directLevelKey[0]] = [];
+                }else{
+                    result[directLevelKey[0]] = null;
+                }
+                directLevelKey =[];
+                return result;
+            }
+        }else{
+            var updatedResult = self.deleteProperty(result,directLevelKey[0],result);
+            directLevelKey =[];
+            return updatedResult;
+        }
+    }else{
+        if(removedKeyStack.length > 0){
+            if(result !== undefined){
+                if(result[childLevelObj[0]] === undefined){
+                    var path = this.options.path.split('.');
+                    if(path.length === 2){
+                      delFlag = false;
+                        for(var j = 0 ; j < removedKeyStack.length; j++){
+                            var source = removedKeyStack[j].split('.');
+                            if(source[0] === 'directChild'){
+                                delete result[source[source.length - 1]];
+                            }else{
+                                currentObjStack = undefined;
+                                arrayObjStack = undefined;
+                                var objectDelete = !!JSON.parse(String(source[3]).toLowerCase());
+                                console.log('Hi i am in stack');
+                                result = self.removeKeyFromStack(result,source[0],source[source.length - 1],source[1],source[2],objectDelete);
+                            }
+                        }
+                   }
+               }
+           }
+         }
+        return result;
+    }
+  },
+  deleteProperty :function(model,delPro,oldModel,Node,next) {
+      var self = this,currentNode = Node;
+      for (var i in model) {
+         if(next === undefined){
+                  currentNode = i;
+          }
+           if(!delFlag){
+              if (typeof model[i] === 'object' && model[i] !== null && model[i].constructor !== Array) {
+                  if(parentLevelObj === i){
+                      delFlag = true;
+                  }
+                  currentObj = i;
+                  if(objectDelete && i === delPro && currentNode === arrayParentKey){
+                       delete model[i];
+                  }else{
+                   self.deleteProperty(model[i],delPro,oldModel,currentNode,'next');
+                  }
+              }else {
+                  if(model[i] !== undefined && model[i] !== null){
+                  if(model[i].constructor === Array){
+                      if((typeof model[i][0] === 'object' || typeof model[i][model[i].length - 1] === 'object') && (!isNumber(delPro) || i == arrayParentKey)){
+                          if(objectDelete && typeof model[i][delPro] === 'object'){
+                              delete model[i][delPro];
+                              model[i][delPro] = '';
+                           }else{
+                              for(var k = 0; k < model[i].length; k++){
+                                  if(parentLevelObj !== undefined && isNumber(parentLevelObj)){
+                                      var key = undefined;
+                                      if(keyHierarchy.indexOf(i) != -1){
+                                          var index = keyHierarchy.indexOf(i) + 1;
+                                          var key = keyHierarchy[index];
+                                      }
+                                      if(k == key){
+                                          self.deleteProperty(model[i][k],delPro,oldModel,currentNode,'next');  
+                                      }
+                                  }else{
+                                      self.deleteProperty(model[i][k],delPro,oldModel,currentNode,'next'); 
+                                  }
+                              } 
+                          }
+                       }else if(typeof model[i][0] !== 'object'){
+                          if(isNumber(delPro) && i === parentLevelObj){
+                          delete  model[i][delPro];
+                          model[i] = model[i].filter(function(n){ return n != undefined }); 
+                          }
+                      }else if(typeof model[i][0] === 'object' && isNumber(delPro) && i === parentLevelObj){
+                          delete  model[i][delPro];
+                          model[i] = model[i].filter(function(n){ return n != undefined }); 
+                      }else if((typeof model[i][0] === 'object' && isNumber(delPro)) && (keyHierarchy.indexOf(parentLevelObj) != -1)){
+                          console.log('ssss11');
+                          self.deleteProperty(model[i][0],delPro,oldModel,currentNode,'next');
+                      }
+                  }
+                  if(currentObj === parentLevelObj || isNumber(parentLevelObj)){
+                      if(i === delPro && currentNode === arrayParentKey){
+                          delete model[i];
+                          
+                     }
+                  }
+             }
+           }
+          }else{
+              if(currentObj === parentLevelObj || isNumber(parentLevelObj)){
+                  if(i === delPro ){ 
+                      delete model[i];
+                      delFlag = false;
+                     }
+              } 
+          }
+      }
+      return model;
+  },
+  removeKeyFromStack :function(model,parentLevelObj,delPro,objLevel,rootLevel,objectDelete) {
+      var self = this;
+      for (var i in model) {
+          if(!delFlag){
+              if (typeof model[i] === 'object' && model[i] !== null && model[i].constructor !== Array) {
+                  if(parentLevelObj === i){
+                      delFlag = true;
+                  }
+                  currentObjStack = i;
+                  self.removeKeyFromStack(model[i],parentLevelObj,delPro,objLevel,rootLevel,objectDelete);
+              } else {
+                  if(model[i] !== undefined && model[i] !== null){
+                      if(model[i].constructor === Array){
+                          if((typeof model[i][0] === 'object' || typeof model[i][model[i].length - 1] === 'object') && (!isNumber(delPro) || i == rootLevel)){
+                              for(var k = 0; k < model[i].length; k++){
+                                  if(i == rootLevel){
+                                      if(k == objLevel){ 
+                                          if(model[i][k][delPro] !== undefined){
+                                              arrayObjStack = i;
+                                          }
+                                          self.removeKeyFromStack(model[i][k], parentLevelObj,delPro,objLevel,rootLevel,objectDelete);
+                                       }
+                                      if(objectDelete){
+                                          delete  model[i][delPro]; 
+                                          model[i][delPro] = '';
+                                       }
+                                  }else{
+                                      arrayObjStack = i;
+                                      self.removeKeyFromStack(model[i][k], parentLevelObj,delPro,objLevel,rootLevel,objectDelete);
+                                  }
+                             }
+                           }else if(typeof model[i][0] !== 'object'){
+                              if(isNumber(delPro) && i === parentLevelObj){
+                                 delete  model[i][delPro];
+                                 model[i][delPro] = '';
+                               }
+                          }else if(typeof model[i][0] === 'object' && isNumber(delPro) && i === parentLevelObj){
+                            delete model[i][delPro];
+                             console.log('3'+ model[i][delPro]);
+                          }
+                      }
+                      if(isNumber(parentLevelObj)){
+                          if(arrayObjStack === rootLevel){
+                              if(i === delPro ){   
+                                  delete model[i];
+                                  arrayObjStack = undefined;
+                                 } 
+                          }
+                      }else if(currentObjStack === parentLevelObj){
+                          if(i === delPro ){   
+                              delete model[i];
+                             }
+                      }
+                  }
+               }
+          }else{
+              if(currentObjStack === parentLevelObj || isNumber(parentLevelObj)){
+                  if(i === delPro ){   
+                      delete model[i];
+                      delFlag = false;
+                     }
+              } 
+          }
+      }
+      return model;
   },
   refreshValue: function() {
     this.value = {};
@@ -3026,19 +3542,26 @@
 
     for(var i in this.editors) {
       if(!this.editors.hasOwnProperty(i)) continue;
-      this.value[i] = this.editors[i].getValue();
+      this.value[i] = this.editors[i].getValue(); 
     }
     
     if(this.adding_property) this.refreshAddProperties();
   },
-  refreshAddProperties: function() {
+  refreshAddProperties: function(node) {
     if(this.options.disable_properties || (this.options.disable_properties !== false && this.jsoneditor.options.disable_properties)) {
       this.addproperty_controls.style.display = 'none';
       return;
     }
-
     var can_add = false, can_remove = false, num_props = 0, i, show_modal = false;
-
+    if(objectPropertiesStack.length > 0){
+        for(var k = 0 ; k < objectPropertiesStack.length; k++){
+            if(objectPropertiesStack[k].parentKey === node){
+                delete this.editors[objectPropertiesStack[k].childKey];
+                delete this.cached_editors[objectPropertiesStack[k].childKey];
+                delete this.schema.properties[objectPropertiesStack[k].childKey];
+            }
+        }
+    }
     // Get number of editors
     for(i in this.editors) {
       if(!this.editors.hasOwnProperty(i)) continue;
@@ -3132,10 +3655,6 @@
         self.addObjectProperty(i);
         editor.setValue(value[i],initial);
       }
-      // Otherwise, remove value unless this is the initial set or it's required
-      else if(!initial && !self.isRequired(editor)) {
-        self.removeObjectProperty(i);
-      }
       // Otherwise, set the value to the default
       else {
         editor.setValue(editor.getDefault(),initial);
@@ -3275,19 +3794,119 @@
     this.hide_move_buttons = this.options.disable_array_reorder || this.jsoneditor.options.disable_array_reorder;
     this.hide_add_button = this.options.disable_array_add || this.jsoneditor.options.disable_array_add;
   },
-  build: function() {
+  build: function(trashIconFlag) {
     var self = this;
-
+    directLevelKey = [];
+    objectPropertiesStack =[];
+    currentObj = undefined;
+    currentObjStack = undefined;
+    arrayObjStack = undefined;
     if(!this.options.compact) {
-      this.header = document.createElement('span');
-      this.header.textContent = this.getTitle();
-      this.title = this.theme.getHeader(this.header);
+      this.title = this.theme.getHeader('');
       this.container.appendChild(this.title);
       this.title_controls = this.theme.getHeaderButtonHolder();
       this.title.appendChild(this.title_controls);
+      this.rootheaderText = document.createElement('span');
+      this.rootheaderText.textContent = this.getTitle();
+      this.title.appendChild(this.rootheaderText);
+      var schemaPath = this.container.getAttribute('data-schemapath').split('.');
+      if(schemaPath.length === 3){
+          this.container.firstChild.setAttribute('style','border-bottom:1px solid #eeeeee;');
+      }
+      if(this.schema.type === 'array'){
+          var addItemIcon = document.createElement('button');
+          addItemIcon.type = 'button';
+          addItemIcon.style.marginTop = '1px';
+          addItemIcon.style.marginLeft = '8px';
+          addItemIcon.style.fontSize = '12px';
+          addItemIcon.title = 'Add Item';
+          addItemIcon.className = 'btn btn-mini json-editor-btn-plus';
+          var innerText = document.createElement('i');
+          innerText.className = 'fa fa-plus';
+          addItemIcon.appendChild(innerText);
+          addItemIcon.addEventListener('click',function(e) {
+              clickAddItem = true;
+              addPropertyFlag = true;
+              propertyDeleteCount = 0;
+              e.preventDefault();
+              e.stopPropagation();
+              enableInputField = true;
+              var i = self.rows.length;
+              if(self.row_cache[i]) {
+                self.rows[i] = self.row_cache[i];
+                self.rows[i].setValue(self.rows[i].getDefault());
+                self.rows[i].container.style.display = '';
+                if(self.rows[i].tab) self.rows[i].tab.style.display = '';
+                self.rows[i].register();
+              }
+              else {
+                self.addRow();
+              }
+              self.active_tab = self.rows[i].tab;
+              self.refreshTabs();
+              self.refreshValue();
+              self.onChange(true);
+              enableInputField = false;
+              addPropertyFlag = false;
+              clickAddItem = false;
+              this.parentElement.firstChild.firstChild.setAttribute('style','margin-top:-8px;');
+              $(this.parentElement.parentElement).find('.well-small')[0].setAttribute('style','padding-bottom:0px;display:block;')
+              var parentEle = this.parentElement.parentElement.getAttribute('data-schemapath').split('.');
+              if(parentEle.length == 3){
+                      $(this.parentElement.parentElement).find('.well-small')[0].firstChild.setAttribute('style','display:block;');
+                  }else{
+                    $(this.parentElement.parentElement).find('.well-small')[0].firstChild.setAttribute('style','display:block;margin-left:9px;border-left:1px dotted #e5e5e5;');
+                    $(this.parentElement.parentElement).find('.well-small')[0].firstChild.lastChild.setAttribute('style','position:relative;display:none;')
+              }
+              $($(this.parentElement.parentElement).find('.well-small')[0].firstChild.lastChild).slideDown(1200);
+              self.collapsed = false;
+              this.parentElement.firstChild.firstChild.firstChild.setAttribute('class','fa fa-sort-desc');
+              this.parentElement.firstChild.firstChild.setAttribute('title','Collapse');
+           });
+          this.title.appendChild(addItemIcon);
+        if(!trashIconHide){
+          var trashicon = document.createElement('button');
+          trashicon.type = 'button';
+          trashicon.style.width = '20px'
+          trashicon.style.marginLeft = '10px'
+          trashicon.className = 'btn btn-mini json-editor-btn-trash deleteChildLevel';
+          var innerText = document.createElement('i');
+          innerText.className = 'fa fa-trash-o';
+          trashicon.appendChild(innerText);
+          trashicon.addEventListener('click',function(e) {
+              e.preventDefault();
+              e.stopPropagation();
+              var text = jsoneditor.root.parseJsonKeyUpperToLower(this.parentElement.firstChild.nextElementSibling.textContent);
+              if(requiredKey.indexOf(text) === -1){
+                  var attrValue = this.parentElement.parentElement.getAttribute('data-schemapath').split('.');
+                  attrValue.splice(0,1);
+                  $(this).closest('.row-fluid')[0].remove();
+                  jsoneditor.root.removeObjectProperty(attrValue,true);
+                  jsoneditor.root.onChange(true);
+              }else{
+                  var errorHolder = $("#config-error-msg-container");
+                  errorHolder.empty();
+                  errorHolder.text('We Can Not Delete The Required Filed - '+ this.parentElement.firstChild.nextElementSibling.textContent);
+                  $('.modal-body').scrollTop(0);
+                  var errorContainer = $("#config-error-container");
+                  errorContainer.fadeIn(500);
+                  $("#remove-error-popup").on('click',function(){
+                      $("#config-error-container").css('display','none');
+                  });
+             }
+          });
+          this.title.appendChild(trashicon);
+        }
+      }
       if(this.schema.description) {
-        this.description = this.theme.getDescription(this.schema.description);
-        this.container.appendChild(this.description);
+        //this.description = this.theme.getDescription(this.schema.description);
+        //this.container.appendChild(this.description);
+        this.title.firstChild.nextSibling.setAttribute('title',this.schema.description);
+      }
+      if(this.schema.required == 'required' || this.schema.required == 'true'){
+          var icon = document.createElement('i');
+          icon.className = 'fa fa-asterisk jsonEditor-form-obj-asterisk';
+          this.title.firstChild.nextSibling.appendChild(icon);
       }
       this.error_holder = document.createElement('div');
       this.container.appendChild(this.error_holder);
@@ -3320,7 +3939,7 @@
     }
 
     // Add controls
-    this.addControls();
+    this.addControls(trashIconFlag);
   },
   onChildEditorChange: function(editor) {
     this.refreshValue();
@@ -3384,7 +4003,20 @@
     var item_info = this.getItemInfo(i);
     var schema = this.getItemSchema(i);
     schema = this.jsoneditor.expandRefs(schema);
-    schema.title = item_info.title+' '+(i+1);
+    var parentPath = this.options.path.split('.');
+    var parentName = parentPath[parentPath.length - 1];
+    //Element label changed
+    if(addPropertyFlag){
+        if(this.schema.items !== undefined){
+            if(this.schema.type === 'array' && this.schema.items.properties === undefined){
+                addPropertyFlag = true;
+             }else{
+                 addPropertyFlag = false; 
+             }
+        }
+    }
+    schema.title = '[ '+(i)+' ]';
+    //schema.title = item_info.title+' '+(i+1);
 
     var editor = this.jsoneditor.getEditorClass(schema);
 
@@ -3490,7 +4122,9 @@
   setValue: function(value, initial) {
     // Update the array's value, adding/removing rows when necessary
     value = value || [];
-    
+    if(addPropertyFlag){
+        value = [];
+    }
     if(!(Array.isArray(value))) value = [value];
     
     var serialized = JSON.stringify(value);
@@ -3595,8 +4229,8 @@
         this.delete_last_row_button.style.display = 'none';
         this.remove_all_rows_button.style.display = 'none';
       }
-      else if(this.value.length === 1) {      
-        this.remove_all_rows_button.style.display = 'none';  
+      else if(this.value.length === 1) {
+        this.remove_all_rows_button.style.display = 'none';
 
         // If there are minItems items in the array, or configured to hide the delete_last_row button, hide the delete button beneath the rows
         if(minItems || this.hide_delete_last_row_buttons) {
@@ -3764,9 +4398,8 @@
     if(value) self.rows[i].setValue(value, initial);
     self.refreshTabs();
   },
-  addControls: function() {
+  addControls: function(trashDisable) {
     var self = this;
-    
     this.collapsed = false;
     this.toggle_button = this.getButton('','collapse',this.translate('button_collapse'));
     this.title_controls.appendChild(this.toggle_button);
@@ -3778,21 +4411,109 @@
       if(self.collapsed) {
         self.collapsed = false;
         if(self.panel) self.panel.style.display = '';
-        self.row_holder.style.display = row_holder_display;
+        var displayProp = self.row_holder.getAttribute('style');
+        var ele = self.row_holder.parentElement.parentElement.getAttribute('data-schemapath').split('.');
+        if(ele.length == 3){
+            self.row_holder.setAttribute('style',''+ displayProp +'');
+        }else{
+            self.row_holder.setAttribute('style','border-left:1px dotted #e5e5e5;margin-left:9px;'+ displayProp +'');
+        }
+        $(self.row_holder).slideDown(700);
         if(self.tabs_holder) self.tabs_holder.style.display = '';
         self.controls.style.display = controls_display;
         self.setButtonText(this,'','collapse','Collapse');
+        this.setAttribute('style','margin-top:-8px;');
+        if(self.rows.length === 0) {
+            clickAddItem = true; 
+            addPropertyFlag = true;
+            propertyDeleteCount = 0;
+            enableInputField = true;
+            self.addRow();
+            self.active_tab = self.rows[0].tab;
+            self.refreshTabs();
+            self.refreshValue();
+            self.onChange(true);
+            enableInputField = false;
+            addPropertyFlag = false;
+            clickAddItem = false;
+            var ele = this.parentElement.parentElement.parentElement;
+            $(ele).find('.well-small')[0].setAttribute('style','padding-bottom:0px;display:none;');
+            var parentEle = ele.getAttribute('data-schemapath').split('.');
+            if(parentEle.length == 3){
+                  $(ele).find('.well-small')[0].firstChild.setAttribute('style','display:block;');
+              }else{
+                $(ele).find('.well-small')[0].firstChild.setAttribute('style','display:block;margin-left:9px;border-left:1px dotted #e5e5e5;');
+              }
+            $($(ele).find('.well-small')[0]).slideDown(1200);
+            self.collapsed = false;
+          }
       }
       else {
         self.collapsed = true;
-        self.row_holder.style.display = 'none';
+        $(self.row_holder).slideUp(700);
         if(self.tabs_holder) self.tabs_holder.style.display = 'none';
         self.controls.style.display = 'none';
-        if(self.panel) self.panel.style.display = 'none';
+        if(self.panel) $(self.panel).slideUp(); //self.panel.style.display = 'none';
         self.setButtonText(this,'','expand','Expand');
+        this.setAttribute('style','margin-top:0px;');
       }
     });
-
+    this.title_controls.parentElement.addEventListener('click',function(e) {
+        e.preventDefault();
+        e.stopPropagation();
+        if(self.collapsed) {
+          self.collapsed = false;
+          if(self.panel) self.panel.style.display = '';
+          var displayProp = self.row_holder.getAttribute('style');
+          var ele = self.row_holder.parentElement.parentElement.getAttribute('data-schemapath').split('.');
+          if(ele.length == 3){
+              self.row_holder.setAttribute('style',''+ displayProp +'');
+          }else{
+              self.row_holder.setAttribute('style','border-left:1px dotted #e5e5e5;margin-left:9px;'+ displayProp +'');
+          }
+          $(self.row_holder).slideDown(700);
+          //self.row_holder.style.display = row_holder_display;
+          if(self.tabs_holder) self.tabs_holder.style.display = '';
+          self.controls.style.display = controls_display;
+          self.setButtonText(this.firstChild.firstChild,'','collapse','Collapse');
+          this.firstChild.firstChild.setAttribute('style','margin-top:-8px;');
+        }else {
+          self.collapsed = true;
+          $(self.row_holder).slideUp(700);
+          if(self.tabs_holder) self.tabs_holder.style.display = 'none';
+          self.controls.style.display = 'none';
+          if(self.panel) $(self.panel).slideUp();
+          self.setButtonText(this.firstChild.firstChild,'','expand','Expand');
+          this.firstChild.firstChild.setAttribute('style','margin-top:0px;');
+        }
+        if(self.rows.length === 0) {
+          clickAddItem = true; 
+          addPropertyFlag = true;
+          propertyDeleteCount = 0;
+          enableInputField = true;
+          self.addRow();
+          self.active_tab = self.rows[0].tab;
+          self.refreshTabs();
+          self.refreshValue();
+          self.onChange(true);
+          enableInputField = false;
+          addPropertyFlag = false;
+          clickAddItem = false;
+          var ele = this.parentElement.parentElement.firstChild;
+          $(ele).find('.well-small')[0].setAttribute('style','padding-bottom:0px;display:none;');
+          var parentEle = ele.getAttribute('data-schemapath').split('.');
+          if(parentEle.length == 3){
+                $(ele).find('.well-small')[0].firstChild.setAttribute('style','display:block;');
+            }else{
+              $(ele).find('.well-small')[0].firstChild.setAttribute('style','display:block;margin-left:9px;border-left:1px dotted #e5e5e5;');
+            }
+          $($(ele).find('.well-small')[0]).slideDown(1200);
+          self.collapsed = false;
+        }
+      });
+     if(this.schema.collapse){
+         this.options.collapsed = true;
+     }
     // If it should start collapsed
     if(this.options.collapsed) {
       $trigger(this.toggle_button,'click');
@@ -3807,7 +4528,7 @@
     }
     
     // Add "new row" and "delete last" buttons below editor
-    this.add_row_button = this.getButton(this.getItemTitle(),'add',this.translate('button_add_row_title',[this.getItemTitle()]));
+    this.add_row_button = this.getButton(this.getItemTitle(),'Add',this.translate('button_add_row_title',[this.getItemTitle()]));
     
     this.add_row_button.addEventListener('click',function(e) {
       e.preventDefault();
@@ -3828,8 +4549,6 @@
       self.refreshValue();
       self.onChange(true);
     });
-    self.controls.appendChild(this.add_row_button);
-
     this.delete_last_row_button = this.getButton(this.translate('button_delete_last',[this.getItemTitle()]),'delete',this.translate('button_delete_last_title',[this.getItemTitle()]));
     this.delete_last_row_button.addEventListener('click',function(e) {
       e.preventDefault();
@@ -4438,7 +5157,10 @@
         if(self.keep_values) editor.setValue(current_value,true);
         editor.container.style.display = '';
       }
-      else editor.container.style.display = 'none';
+      else{
+          editor.container.style.display = 'none';
+      }
+      
     });
     self.refreshValue();
     self.refreshHeaderText();
@@ -4547,11 +5269,12 @@
     this.switcher = this.theme.getSwitcher(this.display_text);
     container.appendChild(this.switcher);
     this.switcher.addEventListener('change',function(e) {
+      addPropertyFlag = true;
       e.preventDefault();
       e.stopPropagation();
-
       self.switchEditor(self.display_text.indexOf(this.value));
       self.onChange(true);
+      addPropertyFlag = false;
     });
 
     this.editor_holder = document.createElement('div');
@@ -4944,8 +5667,11 @@
   build: function() {
     var self = this;
     if(!this.options.compact) this.header = this.label = this.theme.getFormInputLabel(this.getTitle());
-    if(this.schema.description) this.description = this.theme.getFormInputDescription(this.schema.description);
-
+    if(this.schema.description !== undefined){
+        if(!this.schema.description){
+            this.description = this.theme.getFormInputDescription(this.schema.description);
+        }
+    }
     if(this.options.compact) this.container.className += ' compact';
 
     this.input = this.theme.getSelectInput(this.enum_options);
@@ -4961,7 +5687,11 @@
       e.stopPropagation();
       self.onInputChange();
     });
-
+    if(this.schema.required == 'required' || this.schema.required == 'true'){
+        var icon = document.createElement('i');
+        icon.className = 'fa fa-asterisk jsonEditor-form-asterisk';
+        this.label.appendChild(icon);
+    }
     this.control = this.theme.getFormControl(this.label, this.input, this.description);
     this.container.appendChild(this.control);
 
@@ -4993,12 +5723,91 @@
       if(this.schema.options && this.schema.options.select2_options) options = $extend(options,this.schema.options.select2_options);
       this.select2 = window.jQuery(this.input).select2(options);
       var self = this;
+      if(this.schema.parentKey !== undefined && this.options.jsoneditor.options.fieldHide){
+          $(this.input).select2().select2('val',this.schema.enum[0]);
+      }
+      if(this.schema.display !== undefined && this.options.jsoneditor.options.fieldHide){
+              var parentVal = this.options.jsoneditor.schema.properties[Object.keys(this.options.jsoneditor.schema.properties)[0]].properties.parent_type.enum;
+              if(Object.keys(this.options.jsoneditor.options.startval).length === 0){
+                  var parentKey = parentVal[0];
+              }else{
+                  var parentKey =  this.options.jsoneditor.options.startval[Object.keys(this.options.jsoneditor.options.startval)[0]]['parent_type'];
+              }
+              var key = parentKey.split('-').join('_');
+              var splitedKey = key.split('_'); var strStack = [];
+              for(var i = 0; i < splitedKey.length; i++){
+                  var captilizeStr = splitedKey[i].charAt(0).toUpperCase() + splitedKey[i].slice(1);
+                  strStack.push(captilizeStr);
+              }
+              var updatedKey = strStack.join(' ');
+              if(this.input.parentElement.parentElement.firstChild.textContent === updatedKey){
+                  $(this.input).select2().select2('val',this.schema.enum[0]);
+                  this.input.parentElement.parentElement.parentElement.setAttribute('style','display:block');
+              }else{
+                  this.input.parentElement.parentElement.parentElement.setAttribute('style','display:none');
+              }
+      }
       this.select2.on('select2-blur',function() {
         self.input.value = self.select2.select2('val');
         self.onInputChange();
       });
       this.select2.on('change',function() {
+        var parentList = [];
+        for(var m=0; m < optionsList.length; m++){
+            parentList.push(optionsList[m][0]);
+        }
         self.input.value = self.select2.select2('val');
+        if(self.input.value == 'undefined'){
+            if(self.input.parentElement.parentElement.firstChild.textContent == 'Parent Type'){
+                for(var j=0; j < optionsList.length; j++){
+                    var list = optionsList[j];
+                        for(var k=1; k<list.length; k++){
+                            var domEle = $("[data-schematype='string']");
+                            for(var l=0; l<domEle.length;l++){
+                                var splitedKey = $("[data-schematype='string']")[l].getAttribute('data-schemapath').split('.');
+                                var key = splitedKey[splitedKey.length-1];
+                                if(key == list[k]){
+                                    $("[data-schematype='string']")[l].setAttribute('style','display:none');
+                                }
+                            }
+                     }
+                }
+            }
+            
+        }else if(parentList.indexOf(self.input.value) != -1){
+            for(var j=0; j < optionsList.length; j++){
+                if(self.input.value == optionsList[j][0]){
+                    var list = optionsList[j];
+                    for(var k=1; k<list.length; k++){
+                        var domEle = $("[data-schematype='string']");
+                        for(var l=0; l<domEle.length;l++){
+                            var splitedKey = $("[data-schematype='string']")[l].getAttribute('data-schemapath').split('.');
+                            var key = splitedKey[splitedKey.length-1];
+                            if(key == list[k]){
+                                var val = self.jsoneditor.schema.properties[Object.keys(self.jsoneditor.schema.properties)[0]].properties[key].enum;
+                                if(val[0] === ""){
+                                   var keyValue = val[1];
+                                   $($("[data-schematype='string']")[l].firstChild.lastChild.lastChild).select2().select2('val',keyValue);
+                                }
+                                $("[data-schematype='string']")[l].setAttribute('style','display:block');
+                            }
+                        }
+                    }
+                }else{
+                    var list = optionsList[j];
+                    for(var k=1; k<list.length; k++){
+                        var domEle = $("[data-schematype='string']");
+                        for(var l=0; l<domEle.length;l++){
+                            var splitedKey = $("[data-schematype='string']")[l].getAttribute('data-schemapath').split('.');
+                            var key = splitedKey[splitedKey.length-1];
+                            if(key == list[k]){
+                                $("[data-schematype='string']")[l].setAttribute('style','display:none');
+                            }
+                        }
+                    }
+                }
+            }
+        }
         self.onInputChange();
       });
     }
@@ -5289,8 +6098,11 @@
   build: function() {
     var self = this;
     if(!this.options.compact) this.header = this.label = this.theme.getFormInputLabel(this.getTitle());
-    if(this.schema.description) this.description = this.theme.getFormInputDescription(this.schema.description);
-
+    if(this.schema.description !== undefined){
+        if(!this.schema.description){
+            this.description = this.theme.getFormInputDescription(this.schema.description);
+        }
+    }
     if(this.options.compact) this.container.className += ' compact';
 
     this.input = this.theme.getSelectInput(this.enum_options);
@@ -5512,8 +6324,11 @@
   build: function() {
     var self = this, i;
     if(!this.options.compact) this.header = this.label = this.theme.getFormInputLabel(this.getTitle());
-    if(this.schema.description) this.description = this.theme.getFormInputDescription(this.schema.description);
-
+    if(this.schema.description !== undefined){
+        if(!this.schema.description){
+            this.description = this.theme.getFormInputDescription(this.schema.description);
+        }
+    }
     if((!this.schema.format && this.option_keys.length < 8) || this.schema.format === "checkbox") {
       this.input_type = 'checkboxes';
 
@@ -5941,10 +6756,22 @@
     if(!this.options.compact) {
       this.label = this.header = this.theme.getCheckboxLabel(this.getTitle());
     }
-    if(this.schema.description) this.description = this.theme.getFormInputDescription(this.schema.description);
+    if(this.schema.description !== undefined){
+        if(!this.schema.description){
+            this.description = this.theme.getFormInputDescription(this.schema.description);
+        }
+    }
     if(this.options.compact) this.container.className += ' compact';
 
     this.input = this.theme.getCheckbox();
+    if(this.schema.description){
+        this.label.setAttribute('title',this.schema.description);
+    }
+    if(this.schema.required == 'required' || this.schema.required == 'true'){
+        var icon = document.createElement('i');
+        icon.className = 'fa fa-asterisk jsonEditor-form-asterisk';
+        this.label.appendChild(icon);
+    }
     this.control = this.theme.getFormControl(this.label, this.input, this.description);
 
     if(this.schema.readOnly || this.schema.readonly) {
@@ -6213,10 +7040,12 @@
     titles = titles || [];
     select.innerHTML = '';
     for(var i=0; i<options.length; i++) {
-      var option = document.createElement('option');
-      option.setAttribute('value',options[i]);
-      option.textContent = titles[i] || options[i];
-      select.appendChild(option);
+     if(options[i] != ''){
+         var option = document.createElement('option');
+         option.setAttribute('value',options[i]);
+         option.textContent = titles[i] || options[i];
+         select.appendChild(option);
+     }
     }
   },
   getTextareaInput: function() {
@@ -6296,7 +7125,9 @@
       button.appendChild(icon);
       button.innerHTML += ' ';
     }
+    if(text !=='item'){
     button.appendChild(document.createTextNode(text));
+    }
     if(title) button.setAttribute('title',title);
   },
   getTable: function() {
@@ -6446,17 +7277,18 @@
     return el;
   },
   setGridColumnSize: function(el,size) {
-    el.className = 'span'+size;
+    el.className = 'col-xs-'+size;
   },
   getSelectInput: function(options) {
     var input = this._super(options);
-    input.style.width = 'auto';
-    input.style.maxWidth = '98%';
+    //input.style.width = 'auto';
+    //input.style.maxWidth = '98%';
     return input;
   },
   getFormInputField: function(type) {
     var el = this._super(type);
-    el.style.width = '98%';
+    el.style.width = '95%';
+    el.style.height = '28px';
     return el;
   },
   afterInputReady: function(input) {
@@ -6483,7 +7315,9 @@
     el.textContent = text;
     return el;
   },
-  getFormControl: function(label, input, description) {
+  getFormControl: function(label, input, description, trashFlag) {
+    var self = this;
+    //arrayDelFlag = false;
     var ret = document.createElement('div');
     ret.className = 'control-group';
 
@@ -6491,11 +7325,51 @@
     controls.className = 'controls';
 
     if(label && input.getAttribute('type') === 'checkbox') {
-      ret.appendChild(controls);
-      label.className += ' checkbox';
-      label.appendChild(input);
-      controls.appendChild(label);
-      controls.style.height = '30px';
+          ret.appendChild(controls);
+          label.className += ' checkbox';
+          //input.appendChild(label);
+          controls.appendChild(input);
+          controls.appendChild(label);
+          controls.style.height = '30px';
+          if(disableCheckBoxTrash){
+              //if(blockFirstOptions){
+                  var trashButton = document.createElement('button');
+                  trashButton.type = 'button';
+                  trashButton.style.width = '20px'
+                  trashButton.className = 'btn btn-mini json-editor-btn-trash deleteChildLevel';
+                  var iconInnerText = document.createElement('i');
+                  iconInnerText.className = 'fa fa-trash-o';
+                  trashButton.appendChild(iconInnerText);
+                  trashButton.addEventListener('click',function(e) {
+                      e.preventDefault();
+                      e.stopPropagation();
+                      var text = jsoneditor.root.parseJsonKeyUpperToLower(this.parentElement.firstChild.nextElementSibling.textContent);
+                      if(requiredKey.indexOf(text) === -1){
+                          var objPath = [];
+                          var attrValue = this.previousSibling.previousSibling.getAttribute('name').split('[');
+                          attrValue.splice(0,1);
+                          for(var j = 0; j < attrValue.length; j++){
+                               var obj = attrValue[j].substring(0,attrValue[j].length-1);
+                               objPath.push(obj);
+                          }
+                          $(this).closest('.row-fluid')[0].remove();
+                          jsoneditor.root.removeObjectProperty(objPath,true);
+                          jsoneditor.root.onChange(true);
+                      }else{
+                          var errorHolder = $("#config-error-msg-container");
+                          errorHolder.empty();
+                          errorHolder.text('We Can Not Delete The Required Filed - '+ this.parentElement.firstChild.nextElementSibling.textContent);
+                          $('.modal-body').scrollTop(0);
+                          var errorContainer = $("#config-error-container");
+                          errorContainer.fadeIn(500);
+                          $("#remove-error-popup").on('click',function(){
+                              $("#config-error-container").css('display','none');
+                          });
+                      }
+                    });
+                  controls.appendChild(trashButton);
+                  //}
+              }
     }
     else {
       if(label) {
@@ -6503,7 +7377,52 @@
         ret.appendChild(label);
       }
       controls.appendChild(input);
-      ret.appendChild(controls);
+     if(trashFlag){
+         //if(blockFirstOptions){
+          var trashButton = document.createElement('button');
+          trashButton.type = 'button';
+          trashButton.style.width = '20px'
+          trashButton.className = 'btn btn-mini json-editor-btn-trash deleteChildLevel';
+          var iconInnerText = document.createElement('i');
+          iconInnerText.className = 'fa fa-trash-o';
+          trashButton.appendChild(iconInnerText);
+          trashButton.addEventListener('click',function(e) {
+              e.preventDefault();
+              e.stopPropagation();
+              var text = jsoneditor.root.parseJsonKeyUpperToLower(this.parentElement.parentElement.firstChild.textContent);
+              if(requiredKey.indexOf(text) === -1){
+                  var parentPath = this.parentElement.parentElement.parentElement.parentElement.parentElement.parentElement.getAttribute('data-schematype');
+                  var objPath = [];
+                  var attrValue = this.previousSibling.getAttribute('name').split('[');
+                  attrValue.splice(0,1);
+                  for(var j = 0; j < attrValue.length; j++){
+                       var obj = attrValue[j].substring(0,attrValue[j].length-1);
+                       objPath.push(obj);
+                  }
+                  if(parentPath == 'array'){
+                      this.parentElement.parentElement.parentElement.remove();
+                      //arrayDelFlag = true;
+                  }else{
+                      $(this).closest('.row-fluid')[0].remove();
+                  }
+                  jsoneditor.root.removeObjectProperty(objPath,true);
+                  jsoneditor.root.onChange(true);
+              }else{
+                  var errorHolder = $("#config-error-msg-container");
+                  errorHolder.empty();
+                  errorHolder.text('We Can Not Delete The Required Filed - '+ this.parentElement.parentElement.firstChild.textContent);
+                  $('.modal-body').scrollTop(0);
+                  var errorContainer = $("#config-error-container");
+                  errorContainer.fadeIn(500);
+                  $("#remove-error-popup").on('click',function(){
+                      $("#config-error-container").css('display','none');
+                  });
+              }
+            });
+          controls.appendChild(trashButton);
+          //}
+       }
+     ret.appendChild(controls);
     }
 
     if(description) controls.appendChild(description);
@@ -6512,7 +7431,7 @@
   },
   getHeaderButtonHolder: function() {
     var el = this.getButtonHolder();
-    el.style.marginLeft = '10px';
+    //el.style.marginLeft = '10px';
     return el;
   },
   getButtonHolder: function() {
@@ -6522,7 +7441,7 @@
   },
   getButton: function(text, icon, title) {
     var el =  this._super(text, icon, title);
-    el.className += ' btn btn-default';
+    el.className += ' btn btn-mini';
     return el;
   },
   getTable: function() {
@@ -7394,26 +8313,26 @@
 
 JSONEditor.defaults.iconlibs.bootstrap2 = JSONEditor.AbstractIconLib.extend({
   mapping: {
-    collapse: 'chevron-down',
-    expand: 'chevron-up',
+    collapse: 'sort-desc',
+    expand: 'caret-right',
     "delete": 'trash',
     edit: 'pencil',
-    add: 'plus',
+    add: 'plus-sign',
     cancel: 'ban-circle',
     save: 'ok',
     moveup: 'arrow-up',
     movedown: 'arrow-down'
   },
-  icon_prefix: 'icon-'
+  icon_prefix: 'fa fa-'
 });
 
 JSONEditor.defaults.iconlibs.bootstrap3 = JSONEditor.AbstractIconLib.extend({
   mapping: {
-    collapse: 'chevron-down',
-    expand: 'chevron-right',
+    collapse: 'sort-down',
+    expand: 'caret-right',
     "delete": 'remove',
     edit: 'pencil',
-    add: 'plus',
+    add: 'plus-sign',
     cancel: 'floppy-remove',
     save: 'floppy-saved',
     moveup: 'arrow-up',
@@ -7424,26 +8343,26 @@
 
 JSONEditor.defaults.iconlibs.fontawesome3 = JSONEditor.AbstractIconLib.extend({
   mapping: {
-    collapse: 'chevron-down',
-    expand: 'chevron-right',
+    collapse: 'sort-desc',
+    expand: 'caret-right',
     "delete": 'remove',
     edit: 'pencil',
-    add: 'plus',
+    add: 'plus-sign',
     cancel: 'ban-circle',
     save: 'save',
     moveup: 'arrow-up',
     movedown: 'arrow-down'
   },
-  icon_prefix: 'icon-'
+  icon_prefix: 'fa fa-'
 });
 
 JSONEditor.defaults.iconlibs.fontawesome4 = JSONEditor.AbstractIconLib.extend({
   mapping: {
-    collapse: 'caret-square-o-down',
+    collapse: 'sort-desc',
     expand: 'caret-square-o-right',
     "delete": 'times',
     edit: 'pencil',
-    add: 'plus',
+    add: 'plus-sign',
     cancel: 'ban',
     save: 'save',
     moveup: 'arrow-up',
@@ -8007,4 +8926,4 @@
   window.JSONEditor = JSONEditor;
 })();
 
-//# sourceMappingURL=jsoneditor.js.map
\ No newline at end of file
+//# sourceMappingURL=jsoneditor.js.map
