diff -Naur old/contrail-webui-third-party/jnpr_slickgrid/slick.dataview.js new/contrail-webui-third-party/jnpr_slickgrid/slick.dataview.js
--- old/contrail-webui-third-party/jnpr_slickgrid/slick.dataview.js	2014-09-03 11:57:31.000000000 -0700
+++ new/contrail-webui-third-party/jnpr_slickgrid/slick.dataview.js	2014-09-03 11:58:42.000000000 -0700
@@ -30,7 +30,7 @@
 
 
     // private
-    var idProperty = "id";  // property holding a unique row id
+    var idProperty = "cgrid";  // property holding a unique row id
     var items = [];         // data by index
     var rows = [];          // data by row
     var idxById = {};       // indexes by id
diff -Naur old/contrail-webui-third-party/jnpr_slickgrid/slick.grid.css new/contrail-webui-third-party/jnpr_slickgrid/slick.grid.css
--- old/contrail-webui-third-party/jnpr_slickgrid/slick.grid.css	2014-09-03 11:57:31.000000000 -0700
+++ new/contrail-webui-third-party/jnpr_slickgrid/slick.grid.css	2014-09-03 12:08:05.000000000 -0700
@@ -81,7 +81,7 @@
 }
 
 .slick-row.ui-widget-content, .slick-row.ui-state-active {
-  position: absolute;
+/*   position: absolute; */
   border: 0px;
   width: 100%;
 }
diff -Naur old/contrail-webui-third-party/jnpr_slickgrid/slick.grid.js new/contrail-webui-third-party/jnpr_slickgrid/slick.grid.js
--- old/contrail-webui-third-party/jnpr_slickgrid/slick.grid.js	2014-09-03 11:57:31.000000000 -0700
+++ new/contrail-webui-third-party/jnpr_slickgrid/slick.grid.js	2014-09-03 12:06:27.000000000 -0700
@@ -924,7 +924,7 @@
         "." + uid + " .slick-header-column { left: 1000px; }",
         "." + uid + " .slick-top-panel { height:" + options.topPanelHeight + "px; }",
         "." + uid + " .slick-headerrow-columns { height:" + options.headerRowHeight + "px; }",
-        "." + uid + " .slick-cell { height:" + rowHeight + "px; }",
+//        "." + uid + " .slick-cell { height:" + rowHeight + "px; }",
         "." + uid + " .slick-row { height:" + options.rowHeight + "px; }"
       ];
 
@@ -1418,7 +1418,7 @@
         rowCss += " " + metadata.cssClasses;
       }
 
-      stringArray.push("<div class='ui-widget-content " + rowCss + "' style='top:" + getRowTop(row) + "px'>");
+      stringArray.push("<div class='ui-widget-content " + rowCss + " slick-row-master slick_row_" + d.cgrid + "' data-cgrid='" + d.cgrid + "'>");
 
       var colspan, m;
       for (var i = 0, ii = columns.length; i < ii; i++) {
@@ -1664,7 +1664,7 @@
       }
 
       if (h !== oldH) {
-        $canvas.css("height", h);
+//        $canvas.css("height", h);
         scrollTop = $viewport[0].scrollTop;
       }
 
@@ -2272,7 +2272,7 @@
         // if this click resulted in some cell child node getting focus,
         // don't steal it back - keyboard events will still bubble up
         // IE9+ seems to default DIVs to tabIndex=0 instead of -1, so check for cell clicks directly.
-        if (e.target != document.activeElement || $(e.target).hasClass("slick-cell")) {
+    	  if (options.editable && (e.target != document.activeElement || $(e.target).hasClass("slick-cell"))) {
           setFocus();
         }
       }
