diff -Naur old/SlickGrid-master/slick.dataview.js new/SlickGrid-master/slick.dataview.js
--- old/SlickGrid-master/slick.dataview.js	2014-05-05 13:33:22.000000000 +0530
+++ new/SlickGrid-master/slick.dataview.js	2014-09-02 20:07:31.000000000 +0530
@@ -76,6 +76,7 @@
     var onRowCountChanged = new Slick.Event();
     var onRowsChanged = new Slick.Event();
     var onPagingInfoChanged = new Slick.Event();
+    var onDataUpdate = new Slick.Event();
 
     options = $.extend(true, {}, defaults, options);
 
@@ -842,7 +843,12 @@
         pagenum = Math.max(0, Math.ceil(totalRows / pagesize) - 1);
         diff = recalc(items, filter);
       }
-
+      /*onDataUpdate is the new event added so that it will be triggered in all following cases
+      /*    1)RowsChanges
+      /*    2)RowCount changes
+      /*    3)PagingInfoChanges
+      */ 
+      onDataUpdate.notify({previous: countBefore, current: rows.length,rows: diff})
       updated = null;
       prevRefreshHints = refreshHints;
       refreshHints = {};
@@ -1018,7 +1024,8 @@
       // events
       "onRowCountChanged": onRowCountChanged,
       "onRowsChanged": onRowsChanged,
-      "onPagingInfoChanged": onPagingInfoChanged
+      "onPagingInfoChanged": onPagingInfoChanged,
+      "onDataUpdate": onDataUpdate
     });
   }
 
diff -Naur old/SlickGrid-master/slick.grid.css new/SlickGrid-master/slick.grid.css
--- old/SlickGrid-master/slick.grid.css	2014-05-05 13:33:22.000000000 +0530
+++ new/SlickGrid-master/slick.grid.css	2014-09-02 11:19:26.000000000 +0530
@@ -81,7 +81,7 @@
 }
 
 .slick-row.ui-widget-content, .slick-row.ui-state-active {
-  position: absolute;
+/*   position: absolute; */
   border: 0px;
   width: 100%;
 }
diff -Naur old/SlickGrid-master/slick.grid.js new/SlickGrid-master/slick.grid.js
--- old/SlickGrid-master/slick.grid.js	2014-05-05 13:33:22.000000000 +0530
+++ new/SlickGrid-master/slick.grid.js	2014-09-02 11:19:26.000000000 +0530
@@ -924,7 +924,7 @@
         "." + uid + " .slick-header-column { left: 1000px; }",
         "." + uid + " .slick-top-panel { height:" + options.topPanelHeight + "px; }",
         "." + uid + " .slick-headerrow-columns { height:" + options.headerRowHeight + "px; }",
-        "." + uid + " .slick-cell { height:" + rowHeight + "px; }",
+//        "." + uid + " .slick-cell { height:" + rowHeight + "px; }",
         "." + uid + " .slick-row { height:" + options.rowHeight + "px; }"
       ];
 
@@ -1418,7 +1418,7 @@
         rowCss += " " + metadata.cssClasses;
       }
 
-      stringArray.push("<div class='ui-widget-content " + rowCss + "' style='top:" + getRowTop(row) + "px'>");
+      stringArray.push("<div class='ui-widget-content " + rowCss + " slick-row-master slick_row_" + d.id + "' data-id='" + d.id + "'>");
 
       var colspan, m;
       for (var i = 0, ii = columns.length; i < ii; i++) {
@@ -1664,7 +1664,7 @@
       }
 
       if (h !== oldH) {
-        $canvas.css("height", h);
+//        $canvas.css("height", h);
         scrollTop = $viewport[0].scrollTop;
       }
 
@@ -2272,7 +2272,7 @@
         // if this click resulted in some cell child node getting focus,
         // don't steal it back - keyboard events will still bubble up
         // IE9+ seems to default DIVs to tabIndex=0 instead of -1, so check for cell clicks directly.
-        if (e.target != document.activeElement || $(e.target).hasClass("slick-cell")) {
+    	  if (options.editable && (e.target != document.activeElement || $(e.target).hasClass("slick-cell"))) {
           setFocus();
         }
       }
