--- node_modules/kue/lib/http/index.js	2017-03-15 11:48:36.000000000 +0530
+++ node_modules/kue/lib/http/index.js	2017-04-17 12:40:42.000000000 +0530
@@ -13,13 +13,14 @@
 var fs = require('fs');
 var path = require('path');
 var corePath = process.mainModule.exports['corePath'];
-var contrailConfig = process.mainModule.exports.config;
-var userList = contrailConfig.staticAuth;
-
+var configUtils = require(corePath + '/src/serverroot/common/config.utils');
+var config = configUtils.getConfig();
+var userList = config.staticAuth;
 var keyFile = './keys/cs-key.pem';
 var certFile = './keys/cs-cert.pem';
-if (contrailConfig.server_options) {
-    keyFile = contrailConfig.server_options.key_file;
+
+if (config.server_options) {
+    keyFile = config.server_options.key_file;
     if (null != keyFile) {
         keyFile = path.normalize(keyFile);
         if (false == fs.existsSync(keyFile)) {
@@ -28,7 +29,7 @@
     } else {
         keyFile = './keys/cs-key.pem';
     }
-    certFile = contrailConfig.server_options.cert_file;
+    certFile = config.server_options.cert_file;
     if (null != certFile) {
         certFile = path.normalize(certFile);
         if (false == fs.existsSync(certFile)) {
