--- old/node_modules/node-dir/lib/readfiles.js	2014-07-15 22:08:05.000000000 +0530
+++ new/node_modules/node-dir/lib/readfiles.js	2014-07-16 13:49:06.000000000 +0530
@@ -66,13 +66,21 @@
             if (!filename) return done(null, files);
             var file = path.join(dir, filename);
             fs.stat(file, function(err, stat) {
-                if (err) return done(err);
+                if (err) {
+                    next();
+                    return;
+                }
+                //if (err) return done(err);
                 if (stat && stat.isDirectory()) {
                     if (options.recursive) {
                         if (options.matchDir && !options.matchDir.test(filename)) return next();
                         if (options.excludeDir && options.excludeDir.test(filename)) return next();
                         readFiles(file, options, callback, function(err, sfiles) {
-                            if (err) return done(err);
+                            if (err) {
+                                next();
+                                return;
+                            }
+                            //if (err) return done(err);
                             files = files.concat(sfiles);
                             next();
                         });
@@ -84,7 +92,11 @@
                     if (options.shortName) files.push(filename);
                     else files.push(file);
                     fs.readFile(file, options.encoding, function(err, data) {
-                        if (err) return done(err);
+                        //if (err) return done(err);
+                        if (err) {
+                            next();
+                            return;
+                        }
                         if (callback.length > 3)
                             if (options.shortName) callback(null, data, filename, next);
                             else callback(null, data, file, next);
@@ -96,4 +108,4 @@
 
     });
 }
-module.exports = readFiles;
\ No newline at end of file
+module.exports = readFiles;
