--- node_modules/redis/index.js	2017-03-15 11:48:42.000000000 +0530
+++ node_modules/redis/index.js	2017-04-17 12:44:05.000000000 +0530
@@ -6,7 +6,8 @@
     to_array = require("./lib/to_array"),
     events = require("events"),
     crypto = require("crypto"),
-    contrailConfig = process.mainModule.exports["config"],
+    configUtils = require(process.mainModule.exports['corePath'] +
+            '/src/serverroot/common/config.utils'),
     parsers = [], commands,
     connection_id = 0,
     default_port = 6379,
@@ -1128,15 +1129,16 @@
 exports.createClient = function (port_arg, host_arg, options) {
     var port = port_arg || default_port,
         host = host_arg || default_host,
-        redis_client, net_client;
+        redis_client, net_client,
+        config = configUtils.getConfig();
 
     net_client = net.createConnection(port, host);
 
-    if (undefined !== contrailConfig.redis_password) {
+    if (undefined !== config.redis_password) {
         if (undefined === options) {
             options = {};
         }
-        options['auth_pass'] = contrailConfig.redis_password;
+        options['auth_pass'] = config.redis_password;
     }
 
     redis_client = new RedisClient(net_client, options);
